src/index.ts

```diff
@@ -1,49 +1,148 @@
+  );
 }
 
-export class MyStrategy<User> extends Strategy<User, MyStrategyVerifyParams> {
-  name = "change-me";
+export class SiweStrategy<User> extends Strategy<User, VerifierFn> {
+  name = "siwe";
+
+  private options: StrategyOptions;
+  protected verify: StrategyVerifyCallback<User, VerifierFn>;
 
   constructor(
-    options: MyStrategyOptions,
-    verify: StrategyVerifyCallback<User, MyStrategyVerifyParams>
+    options: StrategyOptions,
+    verify: StrategyVerifyCallback<User, VerifierFn>
   ) {
     super(verify);
-    // do something with the options here
+
+    if (!isStrategyOptions(options)) {
+      throw new Error("invalid options object");
+    }
+
+    if (typeof verify !== "function") {
+      throw new Error("verify is not a function");
+    }
+
+    this.options = options;
+    this.verify = verify;
   }
 
   async authenticate(
     request: Request,
     sessionStorage: SessionStorage,
     options: AuthenticateOptions
   ): Promise<User> {
-    return await this.failure(
-      "Implement me!",
-      request,
-      sessionStorage,
-      options
-    );
-    // Uncomment me to do a success response
-    // this.success({} as User, request, sessionStorage, options);
+    const formData = await request.formData();
+
+    const message = formData.get("message");
+    if (typeof message !== "string") {
+      throw new Error(`request formData "message" is not a string`);
+    }
+    const signature = formData.get("signature");
+    if (typeof signature !== "string") {
+      throw new Error(`request formData "signature" is not a string`);
+    }
+
+    let siweMessage: SiweMessage;
+    try {
+      siweMessage = new SiweMessage(message);
+    } catch (err) {
+      // SiweMessage constructor throws a SiweError an object-type message is invalid
+      // https://github.com/spruceid/siwe/blob/23f7e17163ea15456b4afed3c28fb091b39feee3/packages/siwe/lib/client.ts#L352
+      if (err instanceof SiweError) {
+        return await this.failure(err.type, request, sessionStorage, options);
+      }
+
+      // SiweMessage constructor throws a generic Error if string message syntax is invalid
+      // https://github.com/spruceid/siwe/blob/23f7e17163ea15456b4afed3c28fb091b39feee3/packages/siwe-parser/lib/abnf.ts#L327
+      // the string message is parsed using the grammar defined in here:
+      // https://github.com/spruceid/siwe/blob/23f7e17163ea15456b4afed3c28fb091b39feee3/packages/siwe-parser/lib/abnf.ts#L23)
+      if (err instanceof Error) {
+        throw new Error(err.message);
+      }
+
+      throw new Error(
+        `SiweMessage constructor threw an unexpected error: ${String(err)}`
+      );
+    }
+    let user = {} as User;
+    await siweMessage
+      .verify(
+        {
+          signature,
+          domain: this.options.domain,
+        },
+        {
+          provider: this.options.provider,
+          suppressExceptions: true,
+        }
+      )
+      .then(async (siweResponse) => {
```

Comment:
I think we can convert this into a `try { … await siweMessage.verify(); … } catch (err) { … }` structure (`passport-siwe` used `then` because passport strategies don't allow for async `authenticate` functions).